File: frontend/src/types/subnet.ts
import type { Environment } from './ipRecord';

export interface Subnet {
  id: string;
  cidr: string;
  name: string;
  description: string | null;
  gateway: string | null;
  vlan_id: number | null;
  environment: Environment;
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export interface SubnetDetail extends Subnet {
  total_ips: number;
  used_ips: number;
  free_ips: number;
  reserved_ips: number;
}

export interface SubnetCreate {
  cidr: string;
  name: string;
  description?: string;
  gateway?: string;
  vlan_id?: number;
  environment: Environment;
}

export interface SubnetUpdate {
  name?: string;
  description?: string;
  gateway?: string;
  vlan_id?: number;
  environment?: Environment;
}

File: backend/app/schemas/subnet.py
import ipaddress
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field, field_validator

from app.models.ip_record import Environment


class SubnetCreate(BaseModel):
    cidr: str = Field(..., description="Network in CIDR notation, e.g. 192.168.1.0/24")
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    gateway: Optional[str] = Field(None, description="Gateway IPv4 address within the subnet")
    vlan_id: Optional[int] = Field(None, ge=1, le=4094)
    environment: Environment

    @field_validator("cidr")
    @classmethod
    def validate_cidr(cls, v: str) -> str:
        try:
            ipaddress.ip_network(v, strict=False)
        except ValueError as exc:
            raise ValueError(f"Invalid CIDR notation: {v}") from exc
        return v

    @field_validator("gateway")
    @classmethod
    def validate_gateway(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            try:
                ipaddress.ip_address(v)
            except ValueError as exc:
                raise ValueError(f"Invalid gateway IP address: {v}") from exc
        return v


class SubnetUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    gateway: Optional[str] = None
    vlan_id: Optional[int] = Field(None, ge=1, le=4094)
    environment: Optional[Environment] = None

    @field_validator("gateway")
    @classmethod
    def validate_gateway(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            try:
                ipaddress.ip_address(v)
            except ValueError as exc:
                raise ValueError(f"Invalid gateway IP address: {v}") from exc
        return v


class SubnetResponse(BaseModel):
    id: str
    cidr: str
    name: str
    description: Optional[str]
    gateway: Optional[str]
    vlan_id: Optional[int]
    environment: Environment
    created_at: datetime
    updated_at: datetime
    created_by: str
    updated_by: str


class SubnetDetailResponse(SubnetResponse):
    total_ips: int
    used_ips: int
    free_ips: int
    reserved_ips: int
